src/mpid/ch4/shm/posix/posix_progress.c

```diff
@@ -79,25 +74,35 @@ static int progress_recv(int blocking)
         payload += msg_hdr->am_hdr_sz;
         payload_left -= msg_hdr->am_hdr_sz;
 
-        /* note: setting is_local, is_async to 1, 1 */
-        MPIDIG_global.target_msg_cbs[msg_hdr->handler_id] (msg_hdr->handler_id, am_hdr,
-                                                           NULL, in_total_data_sz, 1, 1, &rreq);
-
-        if (!rreq) {
-            MPIDI_POSIX_eager_recv_commit(&transaction);
-            goto fn_exit;
-        } else if (in_total_data_sz == payload_left) {
-            MPIDIG_recv_copy(p_data, rreq);
-            MPIDIG_REQUEST(rreq, req->target_cmpl_cb) (rreq);
-            MPIDI_POSIX_eager_recv_commit(&transaction);
-            MPIDI_POSIX_EAGER_RECV_COMPLETED_HOOK(rreq);
-            goto fn_exit;
-        } else {
-            /* prepare for asynchronous transfer */
-            MPIDIG_recv_setup(rreq);
-
-            MPIR_Assert(MPIDI_POSIX_global.active_rreq[transaction.src_local_rank] == NULL);
-            MPIDI_POSIX_global.active_rreq[transaction.src_local_rank] = rreq;
+        switch (msg_hdr->am_type) {
+            case MPIDI_POSIX_AM_TYPE__HDR:
+                /* note: setting is_local, is_async to 1, 0 */
+                MPIDIG_global.target_msg_cbs[msg_hdr->handler_id] (msg_hdr->handler_id, am_hdr,
+                                                                   payload, payload_left, 1, 0,
+                                                                   NULL);
+                MPIDI_POSIX_eager_recv_commit(&transaction);
+                goto fn_exit;
+                break;
+            case MPIDI_POSIX_AM_TYPE__SHORT:
+                /* note: setting is_local, is_async to 1, 0 */
+                MPIDIG_global.target_msg_cbs[msg_hdr->handler_id] (msg_hdr->handler_id, am_hdr,
+                                                                   payload, payload_left, 1, 0,
+                                                                   NULL);
+                MPIDI_POSIX_eager_recv_commit(&transaction);
+                goto fn_exit;
+                break;
```

Comment:
Are these two cases identical? Maybe you can combine them.

Reply:
Sure.